pcie: Implement a more refined device scanning procedure

jhedberg opened the pull request:
This set of patches implements and takes into use a more refined scanning procedure than the brute-force approach, where the code has been scanning for all possible combinations of BDF (Bus, Device, Function) values. We can do much better than the brute-force approach by doing more careful inspection of the available endpoints this includes things like:

- Check if host controller 0 (BDF(0, 0, 0)) is a multifunction device. If so, the function numbers are all potential host controllers with corresponding bus numbers
- Look for PCI-to-PCI bridge endpoints, and inspect the "secondary bus number" register to get the bus number of bus behind this bridge.
- Only iterate all possible functions of a device if the multifunction bit is set in the endpoint's header type register.

This set of patches introduces a new `pcie_scan()` API which implements the approach procedure, with a flexible set of options that can be used to control the scanning flow, depending on what kind of scan is needed.

Commit:
pcie: Add helper macro to identify valid identifiers

PCI(e) host controllers behave in different ways (some more buggy than
others) in what value they use to indicate that an endpoint is not
present. In most cases the VID/DID is all ones (PCIE_ID_NONE) but in
others it's all zeroes, and some may even have the VID all zeroes and
the DID all ones, or vice-versa.

Add a macro to easily test for all these possibilities. The "all ones"
and "all zeroes" cases have been verified to exist on actual HW
supported by Zephyr, however the test for the mixed cases is simply
based on what Linux considers valid values (drivers/pci/probe.c in the
Linux kernel tree).

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
pcie: Add API for scanning for available endpoints

This adds a generic API to be used for scanning for available PCI
endpoints. It takes a more detailed approach than the "brute force"
based scanning that's so far been used in Zephyr, buy inspecting the
host controller node and bridge nodes, and only scanning for busses and
devices that are actually expected to exist.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
pcie: Convert PCIe device scanning to use pcie_scan() API

Take advantage of the new PCIe scanning API for doing the initial lookup
of PCIe devices specified for a given board.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
pcie: Convert pcie_bdf_lookup() to use pcie_scan()

Use the commmon scanning framework instead of doing a custom brute-force
scan.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
pcie: shell: Use pcie_scan() for scanning for devices

Use the new pcie_scan() API instead of doing a brute-force scan.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

finikorg commented on include/zephyr/dt-bindings/pcie/pcie.h:
```diff
@@ -35,6 +35,16 @@
 
+/* Helper macro to exclude invalid PCIe identifiers. We should really only
+ * need to look for PCIE_ID_NONE, but because of some broken PCI controllers
+ * we have try to cases where both VID & DID are zero or just one of them is
```
the sentence is rather complex for non English speakers ;)

jhedberg replied:
The biggest problem was that it was in the wrong header file (dt-bindings instead of drivers). That's now fixed. I also clarified the documentation a little bit, but not sure how to make it any more clear than that.

dcpleung commented on drivers/pcie/host/pcie.c:
```diff
@@ -370,6 +370,94 @@ pcie_bdf_t pcie_bdf_lookup(pcie_id_t id)
+				secondary = PCIE_BUS_SECONDARY_NUMBER(num);
+			}
+			__fallthrough;
+		case PCIE_CONF_TYPE_CARDBUS_BRIDGE:
```
Also need `__fallthrough` here.

jhedberg replied:
I just removed the CARDBUS case, since we don't do anything special for it.

FWIW, if it's part of the official coding style to require __fallthrough even when there are no lines between the case statements, it seems pretty wasteful, especially if you have lots of case statement stacked up like that. It also doesn't add much value: the purpose of __fallthrough is to make the intent clear whenever there might be some doubt over it. With nothing in between two consecutive case statements (not even empty lines) there really isn't any doubt over the intent. Anyway, this is a topic independent of this PR.

Commit:
pcie: Deprecate pcie_probe() and pcie_bdf_lookup()

These functions are no longer needed with the centralized framework for
scanning for available PCIe devices and discovering their BDF value
through it.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -370,6 +370,93 @@ pcie_bdf_t pcie_bdf_lookup(pcie_id_t id)
+			break;
+		}
+
+		if (do_cb && !opt->cb(bdf, id, opt->cb_data)) {
```
need to test opt->cb validity prior to calling it no?

jhedberg replied:
I can't really think of any valid use cases for using `pcie_scan()` without a callback, so I think I'd just return `-EINVAL` straight away instead of checking for NULL here.

jhedberg replied:
I've added a CHECKIF() for this

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -370,6 +370,93 @@ pcie_bdf_t pcie_bdf_lookup(pcie_id_t id)
+/* Forward declaration needed since scanning a device may reveal a bridge */
+static bool scan_bus(int bus, const struct pcie_scan_opt *opt);
+
+static bool scan_dev(int bus, int dev, const struct pcie_scan_opt *opt)
```
it's probably really unlikely to happen, but shouldn't you have a call counter here to limit the recursive calls to something "sane" (whatever that could be) ?

jhedberg replied:
You've essentially got some weird or broken HW in that case. Not sure if it's worth it, but a call counter sounds a bit arbitrary. Another option would be to have a bitmask of already processed bus numbers, but that might also be excessive overhead considering that this really indicates that there's something wrong with the HW.

jhedberg replied:
I thought a bit more about this, and I don't think we should bother. In the Linux case it always allocates and programs new bus numbers to found busses (by writing a new number to the register), so it doesn't even try to do some kind of circular dependency kind of detection.

jhedberg commented:
One more commit added to deprecate `pcie_probe()` and `pcie_bdf_lookup()`. Those don't have any in-tree users anymore and are not needed since drivers should rely on the new centralized scan for devices.

jhedberg commented:
Just pushed an update which changes the two bools in the scan_opts to a flags bitmap. This should make the API more easily extensible in the future.

finikorg commented on include/zephyr/drivers/pcie/pcie.h:
```diff
@@ -37,6 +37,16 @@ typedef uint32_t pcie_bdf_t;
  */
 typedef uint32_t pcie_id_t;
 
+/* Helper macro to exclude invalid PCIe identifiers. We should really only
```
Should we document it like other macros?
```
/**
 * @brief ...
```

jhedberg replied:
I think this is really more of an internal API, only needed by `shell.c` and `pcie.c`. We might want to move such things to a new `drivers/pcie/host/pcie_internal.h` file or similar. Anyway, I guess it's a bit out of scope for this PR.

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -370,6 +371,106 @@ pcie_bdf_t pcie_bdf_lookup(pcie_id_t id)
 	return PCIE_BDF_NONE;
 }
 
+static bool scan_flag(const struct pcie_scan_opt *opt, uint32_t flag)
```
should it be inlined? (though the compiler will probably be smart by itself here)

jhedberg replied:
Yeah, I think the sane approach is to trust the compiler, and only if you find it not doing the right thing to override it manually with an inline declaration in the code. In this case I'm almost sure that it'll always be inlined, unless you explicitly disable optimizations.

jhedberg commented:
@dcpleung could you take another look? Thanks.

jhedberg commented:
Had to rebase and force push due to a conflict in the release notes. @finikorg @tbursztyka could you re-approve?
